File: app/src/main/java/com/haishinkit/rtmp/messages/RTMPWindowAcknowledgementSizeMessage.java
package com.haishinkit.rtmp.messages;

import com.haishinkit.rtmp.RTMPChunk;
import com.haishinkit.rtmp.RTMPConnection;
import com.haishinkit.rtmp.RTMPSocket;

import java.nio.ByteBuffer;

public final class RTMPWindowAcknowledgementSizeMessage extends RTMPMessage {
    private static final int CAPACITY = 4;

    private int size = 0;

    public RTMPWindowAcknowledgementSizeMessage() {
        super(Type.ACK);
    }

    public int getSize() {
        return size;
    }

    public RTMPWindowAcknowledgementSizeMessage setSize(final int size) {
        this.size = size;
        return this;
    }

    @Override
    public ByteBuffer encode(final RTMPSocket socket)  {
        ByteBuffer buffer = ByteBuffer.allocate(CAPACITY);
        buffer.putInt(size);
        return buffer;
    }

    @Override
    public RTMPMessage decode(final RTMPSocket socket, final ByteBuffer buffer) {
        return setSize(buffer.getInt());
    }

    @Override
    public RTMPMessage execute(final RTMPConnection connection) {
        connection.getSocket().doOutput(
                RTMPChunk.ZERO,
                new RTMPWindowAcknowledgementSizeMessage()
                        .setSize(size)
                        .setChunkStreamID(RTMPChunk.CONTROL)
        );
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/messages/RtmpWindowAcknowledgementSizeMessage.kt
package com.haishinkit.rtmp.messages

import com.haishinkit.rtmp.RtmpChunk
import com.haishinkit.rtmp.RtmpConnection
import java.nio.ByteBuffer

/**
 * 5.4.4. Window Acknowledgement Size (5)
 */
internal class RtmpWindowAcknowledgementSizeMessage : RtmpMessage(TYPE_ACK) {
    var size: Int = 0
        private set
    override var length: Int = CAPACITY

    override fun encode(buffer: ByteBuffer): RtmpMessage {
        buffer.putInt(size)
        return this
    }

    override fun decode(buffer: ByteBuffer): RtmpMessage {
        size = buffer.int
        return this
    }

    override fun execute(connection: RtmpConnection): RtmpMessage {
        val ack = connection.messageFactory.createRtmpWindowAcknowledgementSizeMessage()
        ack.size = size
        ack.chunkStreamID = RtmpChunk.CONTROL
        connection.doOutput(RtmpChunk.ZERO, ack)
        return this
    }

    companion object {
        private const val CAPACITY = 4
    }
}
